src/dtls13.c

```diff
@@ -0,0 +1,2186 @@
+    word32 msg_length)
+{
+    Dtls13HandshakeHeader *hdr;
+
+    hdr = (Dtls13HandshakeHeader *)output;
+
+    hdr->msg_type = msg_type;
+    c32to24((word32)msg_length, hdr->length);
+    c16toa(ssl->keys.dtls_handshake_number, hdr->messageSeq);
+
+    c32to24(frag_offset, hdr->fragmentOffset);
+    c32to24(frag_length, hdr->fragmentLength);
+
+    return 0;
+}
+
+static int Dtls13TypeIsEncrypted(enum HandShakeType hs_type)
+{
+    switch (hs_type) {
+    case hello_request:
+    case hello_verify_request:
+    case client_hello:
+    case server_hello:
+        return 0;
+    case encrypted_extensions:
+    case session_ticket:
+    case end_of_early_data   :
+    case hello_retry_request :
+    case certificate         :
+    case server_key_exchange :
+    case certificate_request :
+    case server_hello_done   :
+    case certificate_verify  :
+    case client_key_exchange :
+    case finished            :
+    case certificate_status  :
+    case key_update          :
+    case change_cipher_hs    :
+    case message_hash        :
+    case no_shake            :
+        return 1;
+    }
+
+    /* abide compilers */
+    return 0;
+}
+
+static int Dtls13GetRnMask(
+    WOLFSSL *ssl, const byte *ciphertext, byte *mask, enum rnDirection dir)
+{
+    Ciphers *c;
+    int ret;
+
+    if (dir == PROTECT)
+        c = &ssl->dtlsRecordNumberEncrypt;
+    else
+        c = &ssl->dtlsRecordNumberDecrypt;
+
+    if (ssl->specs.bulk_cipher_algorithm == wolfssl_aes_gcm) {
+#ifdef HAVE_AESGCM
+        if (c->aes == NULL)
+            return BAD_STATE_E;
+        return  wc_AesEncryptDirect(c->aes, mask, ciphertext);
+#endif /* HAVE_AESGCM */
+    }
+
+    if (ssl->specs.bulk_cipher_algorithm == wolfssl_chacha) {
+#ifdef HAVE_CHACHA
+        word32 counter;
+
+        if (c->chacha == NULL)
+            return BAD_STATE_E;
+
+        /* assuming CIPHER[0..3] should be interpreted as big endian 32-bits
+           integer. The draft rfc isn't really clear on that. See sec 4.2.3 of
+           the draft. See also Section 2.3 of the Chacha RFC. */
+        ato32(ciphertext, &counter);
+        ret = wc_Chacha_SetIV(c->chacha, &ciphertext[4], counter);
+        if (ret != 0)
+            return ret;
+
+        XMEMSET(mask, 0, DTLS13_RN_MASK_SIZE);
+
+        return wc_Chacha_Process(c->chacha, mask, mask, DTLS13_RN_MASK_SIZE);
+#endif /* HAVE_CHACHA */
+    }
+
+    return NOT_COMPILED_IN;
+}
+
+static int Dtls13EncryptDecryptRecordNumber(WOLFSSL *ssl, byte *seq,
+    int SeqLength, const byte *ciphertext, enum rnDirection dir)
+{
+    byte mask[DTLS13_RN_MASK_SIZE];
+    int ret;
+
+    ret = Dtls13GetRnMask(ssl, ciphertext, mask, dir);
+    if (ret != 0)
+        return ret;
+
+    xorbuf(seq, mask, SeqLength);
+
+    return 0;
+}
+
+static int Dtls13RtxNeedsExplicitAck(WOLFSSL *ssl, enum HandShakeType hs)
+{
+
+    (void)hs;
+
+#ifndef NO_WOLFSSL_SERVER
+    if (ssl->options.side == WOLFSSL_SERVER_END && (hs == finished))
```

Comment:
I'm still not convinced that only `finished` should be marked as needing an explicit ack. This combined with the following section
```
    if (ssl->keys.dtls_peer_handshake_number ==
            ssl->keys.dtls_expected_peer_handshake_number &&
        Dtls13RtxNeedsExplicitAck(ssl, hs))
        fsm->sendAcks = 1;
```
feels like someone might break this in the future because the connection between these two sections is not obvious. I'll still ask you to consider marking all of the client's final flight as needing explicit acks and thinking of another way to not send them all immediately.

Reply:
It may be the naming it's also unfortunate, `Dtls13RtxNeedsExplicitAck`  more than "does the message need an explicit ACK?" means "Is it the last message of a flight that needs an explicit ACK?". 

How do you see moving the `if(Dtls13RtxNeedsExplicitAck) sendAck = 1` check after processing the message? Do you think it would make more sense?

On top of my mind, the only point where we may need to send an ACK immediately for `Cert/CertVerify` is in case the server:
1. receives and buffers the finished message, it sends an ACK for the finished
2. client resends `cert/cert verify`
3. server processes everything, sends the ACKs after processing the finished message buffered  previously
4. The ACK on point 3 is lost
5. client resends `cert/cert verify`

At this point, we detect the `cert/cert verify` as a retransmission. In this case, when the `sendMoreAck` option is on, we ack those messages immediately, otherwise we will send the ack when (fast) timeout expires.

It looks sound to me. Am I missing something?

Reply:
>It may be the naming it's also unfortunate, Dtls13RtxNeedsExplicitAck more than "does the message need an explicit ACK?" means "Is it the last message of a flight that needs an explicit ACK?".

In that case I would recommend changing it to "Dtls13IsLastMsgInFlight" or something so that its more obvious that that is the purpose of this check. Possibly split it into two functions (if needed) where one checks if it needs to be acked and the other if it is the last msg of a flight.

>How do you see moving the if(Dtls13RtxNeedsExplicitAck) sendAck = 1 check after processing the message? Do you think it would make more sense?

I think that this is a better way to go about doing this. Doing the check as soon as the msg arrives forces you to account for network issues which are making everything more complicated.

> [the rest]

I agree that we don't need to send acks immediately for other msgs. In fact we should, as you mentioned, only do it on a fast timeout. I feel like the naming is misleading and the connection with the section of code I mentioned isn't obvious.

Reply:
Hi @julek-wolfssl ,
can you please re-check this with the last version?

Reply:
Please just add a comment to this function that it expects that the entire last flight will be included in the `finished` ACK.